File: maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int ans = nums[0];
        int prevMax = nums[0];
        int prevMin = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];            
            int currMax = Collections.max(List.of(num * prevMax, num * prevMin, num));
            int currMin = Collections.min(List.of(num * prevMax, num * prevMin, num));
            ans = Collections.max(List.of(ans, currMax, currMin));
            prevMax = currMax;
            prevMin = currMin;
        }
        
        return ans;
    }
}

File: maximum-product-subarray/maximum-product-subarray.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_product = nums[0]
        min_product = nums[0]
        ans = nums[0]
        
        for i in range(1, len(nums)):
            num = nums[i]
            temp = max_product
            max_product = max(max_product * num, min_product * num, num)
            min_product = min(temp * num, min_product * num, num)
            ans = max(ans, max_product, min_product)
        
        return ans
        